commands/ls.go

```diff
@@ -68,3 +81,67 @@ func cmdLs(c *cli.Context) {
 
 	w.Flush()
 }
+
+func parseFilters(filters []string) (FilterOptions, error) {
+	options := FilterOptions{}
+	for _, f := range filters {
+		kv := strings.SplitN(f, "=", 2)
+		key, value := kv[0], kv[1]
+
+		switch key {
+		case "swarm":
+			options.SwarmName = append(options.SwarmName, value)
+		case "driver":
+			options.DriverName = append(options.DriverName, value)
+		default:
+			return options, fmt.Errorf("Unsupported filter key '%s'", key)
+		}
+	}
+	return options, nil
+}
+
+func filterHosts(hosts []*libmachine.Host, filters FilterOptions) []*libmachine.Host {
+	filteredHosts := []*libmachine.Host{}
+
+	swarmMasters := getSwarmMasters(hosts)
+
+	for _, h := range hosts {
+		if filterHost(h, filters, swarmMasters) {
+			filteredHosts = append(filteredHosts, h)
+		}
+	}
+	return filteredHosts
+}
+
+func getSwarmMasters(hosts []*libmachine.Host) map[string]string {
+	swarmMasters := make(map[string]string)
+	for _, h := range hosts {
+		swarmOptions := h.HostOptions.SwarmOptions
+		if swarmOptions != nil && swarmOptions.Master {
+			swarmMasters[swarmOptions.Discovery] = h.Name
+		}
+	}
+	return swarmMasters
+}
+
+func filterHost(host *libmachine.Host, filters FilterOptions, swarmMasters map[string]string) bool {
+	if len(filters.SwarmName) > 0 {
```

Comment:
Are these `len()` checks really needed?  If you attempt a `range` over a nil slice, I think it will just skip it.